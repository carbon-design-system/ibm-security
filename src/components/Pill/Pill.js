/**
 * @file Pill.
 * @copyright IBM Security 2019
 */

import PropTypes from 'prop-types';
import classnames from 'classnames';
import React from 'react';

import { Decorator } from '../DataDecorator';

import { getComponentNamespace } from '../../globals/namespace/index';

export const namespace = getComponentNamespace('pill');

/**
 * Pill component.
 */
const Pill = ({ className, inline, type, value }) => {
  const classes = classnames(className, namespace, {
    [`${namespace}--inline`]: inline,
  });

  return (
    <Decorator
      className={classes}
      inline={inline}
      noIcon
      type={type}
      value={value}
    />
  );
};

Pill.propTypes = {
  /** @type {string} Extra classes to add. */
  className: PropTypes.string,

  /** @type {boolean} Whether the Pill should be treated and styled as an inline element. */
  inline: PropTypes.bool,

  /** @type {string} The type of data. */
  type: PropTypes.string.isRequired,

  /** @type {string} The value of the data. */
  value: PropTypes.string.isRequired,
};

Pill.defaultProps = {
  className: '',
  inline: false,
};

export default Pill;
