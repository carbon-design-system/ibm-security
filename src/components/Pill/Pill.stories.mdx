import { Meta, Story, Preview } from '@storybook/addon-docs/blocks';
import { Pill } from '../../';
 
<Meta title="Pillz" component={Pill} />
 
# Pill
 
With `MDX` we can define a story for `Pill` right in the middle of our
markdown documentation.
 
<Preview>
  <Story name="all Pilles">
    <form>
      <Pill value='127.0.0.1' type='IP' />
    </form>
  </Story>
</Preview>