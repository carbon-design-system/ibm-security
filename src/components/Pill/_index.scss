////
/// Pill component.
/// @group pill
/// @copyright IBM Security 2019
////

@import '../Component/mixins';
@import 'mixins';

@import '../DataDecorator/index';

@include security--component($name: pill) {
  @include pill;
}
