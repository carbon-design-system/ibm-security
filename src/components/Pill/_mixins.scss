////
/// Pill mixins.
/// @group pill
/// @copyright IBM Security 2019
////

@import '../../globals/index';

@mixin pill {
  border-radius: carbon--mini-units($count: 12.5);
  padding: 0 $carbon--spacing-03;

  &--inline {
    border-radius: carbon--mini-units($count: 6.25);
    display: inline-flex;
    padding: 0 $carbon--spacing-02;
  }
}
