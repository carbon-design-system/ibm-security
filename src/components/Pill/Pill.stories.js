/**
 * @file Pill stories.
 * @copyright IBM Security 2019
 */

import React from 'react';
import { storiesOf } from '@storybook/react';

import { text } from '@storybook/addon-knobs';

import { components } from '../../../.storybook';

import Pill from './Pill';

storiesOf(components('Pill'), module).add(`default`, () => (
  <div style={{ padding: '64px' }}>
    <Pill value={text('value', '127.0.0.1')} type={text('type', 'IP')} />
    <br />
    <br />
    <p style={{ width: '400px' }} className="bx--type-body-long-01">
      {
        'This is an inline decorator that appears inside a paragraph, alongside text. '
      }
      <Pill
        value={text('value', '127.0.0.1')}
        type={text('type', 'IP')}
        inline
      />
      {
        ' This variation of the decorator component is smaller to match the font size of the paragraph text.'
      }
    </p>
  </div>
));
