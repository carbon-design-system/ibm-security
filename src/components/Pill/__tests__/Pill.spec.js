/**
 * @file Pill tests.
 * @copyright IBM Security 2020 - 2021
 */

import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import { Pill } from '../../..';
import { namespace } from '../Pill';

describe('Pill', () => {
  test('should have no Axe or DAP violations', async () => {
    const { container } = render(<Pill value="127.0.0.1" type="IP" />);
    await expect(container).toHaveNoAxeViolations();
    await expect(container).toHaveNoDAPViolations('Pill');
  });

  test('should add a custom class', () => {
    const { getByText } = render(
      <Pill value="127.0.0.1" type="IP" className="custom-class" />
    );
    expect(getByText(/127.0.0.1/i).parentNode).toHaveClass('custom-class');
  });

  test('should apply `value`', () => {
    const { getByText } = render(<Pill value="127.0.0.1" type="IP" />);
    expect(getByText(/127.0.0.1/i)).toBeVisible();
  });

  test('should apply `type`', () => {
    const { getByText } = render(
      <Pill value="127.0.0.1" type="IP" className="custom-class" />
    );
    expect(getByText(/IP/i)).toBeVisible();
  });

  test('should be inline when `inline` is set to `true`', () => {
    const { container } = render(<Pill value="127.0.0.1" type="IP" inline />);
    expect(container.firstElementChild).toHaveClass(`${namespace}--inline`);
  });

  test('should not have an icon', () => {
    const { container } = render(<Pill value="127.0.0.1" type="IP" />);
    expect(container.querySelector('svg')).toBe(null);
  });

  test('should be inert and non-interactive', () => {
    const { getByText } = render(
      <div>
        <Pill value="127.0.0.1" type="IP" />
        <button>test button</button>
      </div>
    );
    userEvent.tab();
    // Expect only the extra test button to be interactive:
    expect(getByText(/127.0.0.1/i).parentNode).not.toHaveFocus();
    expect(getByText(/test button/i)).toHaveFocus();
  });
});
