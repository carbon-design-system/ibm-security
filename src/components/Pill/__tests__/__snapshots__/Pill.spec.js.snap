// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pill renders correctly 1`] = `
<Decorator
  active={false}
  className="security--pill"
  inert={true}
  inline={false}
  noIcon={true}
  onClick={[Function]}
  scoreThresholds={
    Array [
      0,
      4,
      7,
      10,
    ]
  }
  title=""
  type="IP"
  value="127.0.0.1"
/>
`;
